pkg/supervisor/supervisor.go

```diff
@@ -0,0 +1,247 @@
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	_, exists := s.workers[worker.Name]
+	if exists {
+		panic(fmt.Sprintf("worker already exists: %s", worker.Name))
+	}
+	s.workers[worker.Name] = worker
+	s.names = append(s.names, worker.Name)
+}
+
+func (s *Supervisor) remove(worker *Worker) {
+	delete(s.workers, worker.Name)
+	var newNames []string
+	for _, name := range s.names {
+		if name == worker.Name {
+			continue
+		} else {
+			newNames = append(newNames, name)
+		}
+	}
+	s.names = newNames
+}
+
+// A supervisor will run until all its workers exit. There are
+// multiple ways workers can exit:
+//
+//   - normally (returning a non-nil error)
+//   - error (either via return result or panic)
+//   - graceful shutdown request
+//   - canceled context
+//
+// A normal exit does not trigger any special action other than
+// causing Run to return if it is the last worker.
+//
+// If a worker exits with an error, the behavior depends on the value
+// of the Retry flag on the worker. If Retry is true, the worker will
+// be restarted. If not the supervisor shutdown sequence is triggerd.
+//
+// The supervisor shutdown sequence can be deliberately triggered by
+// invoking supervisor.Shutdown(). This can be done from any goroutine
+// including workers.
+//
+// The graceful shutdown sequence shuts down workers in an order that
+// respects worker dependencies.
+func (s *Supervisor) Run() []error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	for len(s.workers) > 0 {
+		s.reconcile()
+		s.changed.Wait()
+	}
+	return s.errors
+}
+
+// Triggers a graceful shutdown sequence. This can be invoked from any
+// goroutine.
+func (s *Supervisor) Shutdown() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.shuttingDown = true
+	s.changed.Broadcast()
+}
+
+func (s *Supervisor) dependents(worker *Worker) (result []*Worker) {
+	for _, n := range s.names {
+		w := s.workers[n]
+		for _, r := range w.Requires {
+			if r == worker.Name {
+				result = append(result, w)
+				break
+			}
+		}
+	}
+	return
+}
+
+// make sure anything that would like to be running is actually
+// running
+func (s *Supervisor) reconcile() {
+	if s.shuttingDown {
+		s.reconcileShutdown()
+	} else {
+		s.reconcileNormal()
+	}
+}
+
+// shutdown anything that is safe to shutdown and not already shutdown
+func (s *Supervisor) reconcileShutdown() {
+OUTER:
+	for _, n := range s.names {
+		w := s.workers[n]
+		if w.process != nil && !w.process.shutdownClosed {
+			for _, d := range s.dependents(w) {
+				if s.workers[d.Name].process != nil {
+					log.Printf("cannot shutdown %s, %s still running", n, d.Name)
+					continue OUTER
+				}
+			}
+			close(w.process.shutdown)
+			w.process.shutdownClosed = true
+		}
+	}
+}
+
+// launch anything that is safe to launch and not already launched
+func (s *Supervisor) reconcileNormal() {
+OUTER:
+	for _, n := range s.names {
+		w := s.workers[n]
+		if w.process == nil {
+			for _, r := range w.Requires {
+				process := s.workers[r].process
```

Comment:
This seems like it could be an easy accidental panic (de-reference nil), if you forget to .Supervise() a worker that another worker depends on.